### PR TITLE
refactor(rxPaginate): EO-1138 Refactor Tables Pagination

### PR DESCRIPTION
**Pull request by satyamkotakonda**
JIRA: https://jira.rax.io/browse/EO-1138

### LGTMs
- [x] Dev LGTM
- [x] Dev+Vidyo LGTM
- [x] Design LGTM
- [x] QE LGTM

Before
---------

After
--------

### CODE REVIEW

**Comment by satyamkotakonda** [comment]
@CITguy  Story not about  rxCopy  but rxCopy related test cases are failing.

screen shot 2016-09-06 at 2 57 10 pm

File: src/elements/Tables/docs/tables.docs.html [hunk 1]
```diff
@@ -86,8 +86,43 @@
       .pagination-per-page { display: none; }
   }
 </pre>
-<rx-example name="table.paginate"></rx-example>
+<h3>
+  <rx-permalink>UI-Based Pagination</rx-permalink>
+  <a class="button xs inline" href="ngdocs/index.html#/api/elements.directive:rxPaginate/">View API</a>
```
Comment (CITguy) [inline src/elements/Tables/docs/tables.docs.html]: Move the `View API` link to the "Table Pagination" heading.

File: src/elements/Tables/docs/tables.docs.html [hunk 2]
```diff
@@ -86,8 +86,43 @@
...
+<ul class="list">
+  <li>
+    UI-based pagination, where all items are retrieved at once, and paginated in the UI
+  </li>
+</ul>
+<rx-example name="table.paginate.ui"></rx-example>
+
+<h3>
+  <rx-permalink>API-Based Pagination</rx-permalink>
+  <a class="button xs inline" href="ngdocs/index.html#/api/elements.directive:rxPaginate/">View API</a>
```
Comment (CITguy) [inline src/elements/Tables/docs/tables.docs.html]: Move the `View API` link to the "Table Pagination" heading.

File: src/elements/Tables/docs/tables.docs.html [hunk 3]
```diff
@@ -60,7 +60,7 @@
 <rx-example name="table.actionMenu"></rx-example>
 
 
-<h3><rx-permalink>Table Footer Pagination via rxPaginate</rx-permalink></h3>
+<h3><rx-permalink>Table Pagination</rx-permalink></h3>
```
Comment (CITguy) [inline src/elements/Tables/docs/tables.docs.html]: Please change to `<h2>`

File: src/components/rxPaginate/rxPaginate.module.js [hunk 4]
```diff
@@ -1,491 +0,0 @@
-/**
- * @ngdoc overview
- * @name rxPaginate
- * @description
- * # rxPaginate Component
- *
- * The rxPaginate component adds pagination to a table.
- *
- * Two different forms of pagination are supported:
```
Comment (CITguy) [inline src/components/rxPaginate/rxPaginate.module.js]: This documentation needs to be kept. Otherwise, it will be difficult to explain how to implement different forms of pagination.

Please move ngdocs to `src/elements/Tables/scripts/rxPaginate.js`

File: src/elements/Tables/scripts/rxLoadingOverlay.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
-angular.module('encore.ui.rxPaginate')
+angular.module('encore.ui.elements')
 /**
  * @ngdoc directive
  * @name rxPaginate.directive:rxLoadingOverlay
```
Comment (CITguy) [inline src/elements/Tables/scripts/rxLoadingOverlay.js]: `@name` needs to be updated to `elements.directive:rxLoadingOverlay`

File: src/elements/Tables/docs/tables.docs.html [hunk 6]
```diff
@@ -86,8 +86,43 @@
       .pagination-per-page { display: none; }
   }
 </pre>
-<rx-example name="table.paginate"></rx-example>
+<h3>
```
Comment (CITguy) [inline src/elements/Tables/docs/tables.docs.html]: Please add blank line above heading.

File: src/elements/Tables/docs/tables.docs.html [hunk 7]
```diff
@@ -60,7 +60,7 @@
 <rx-example name="table.actionMenu"></rx-example>
 
 
-<h3><rx-permalink>Table Footer Pagination via rxPaginate</rx-permalink></h3>
+<h3><rx-permalink>Table Pagination</rx-permalink></h3>
 <ul class="list">
   <li>
```
Comment (CITguy) [inline src/elements/Tables/docs/tables.docs.html]: Considering that the examples are just below on the same page, this list item is moot. 

Please remove this list item.

**Comment by satyamkotakonda** [comment]
@CITguy  Updated changes and please review.

**Comment by satyamkotakonda** [comment]
@CITguy  

It is giving expected data while testing manually.



It is giving wrong data while running midways.

screen shot 2016-09-12 at 12 58 13 pm


Please help me here.

File: utils/rx-page-objects/test/rxPaginate.midway.js [hunk 8]
```diff
@@ -73,19 +73,14 @@ var table = Page.create({
...
     describe('Non present pagination exercise', encore.exercise.rxPaginate({
         instance: encore.rxPaginate.initialize($('#does-not-exist')),
         isPresent: false
     }));
 
-    describe('Non displayed pagination exercise', encore.exercise.rxPaginate({
-        instance: encore.rxPaginate.initialize($('#rx-paginate-hidden')),
-        isDisplayed: false
-    }));
-
```
Comment (Droogans) [inline utils/rx-page-objects/test/rxPaginate.midway.js]: This was filed under a commit titled "Update midways". That seems a bit inaccurate when you're outright deleting tests.

Why is this removed?

Comment (satyamkotakonda) [inline utils/rx-page-objects/test/rxPaginate.midway.js]: @Droogans  I have removed the  rx-paginate-hidden element as per requirement.
Here is link:

[rx-paginate-hidden](https://github.com/rackerlabs/encore-ui/pull/1795/files#diff-ed6144ded293ff23fae71064b4dd5e35L4)

Comment (Droogans) [inline utils/rx-page-objects/test/rxPaginate.midway.js]: Where is this requirement? The story says nothing about removing lower-level exercises or `rx-debug` markup from the example.

File: utils/rx-page-objects/test/rxPaginate.midway.js [hunk 9]
```diff
@@ -99,12 +95,14 @@ describe('rxPaginate', function () {
...
         beforeEach(function () {
-            encore.rxMisc.scrollToElement(table.tblResults);
             table.textFilter = '';
             table.selectFilter({
                 Os: { All: true }
             });
             nameColumn.sortAscending();
+            encore.rxMisc.scrollToElement(table.tblResults, {
+                positionOnScreen: 'middle'
+            });
```
Comment (CITguy) [inline utils/rx-page-objects/test/rxPaginate.midway.js]: scrolling has to happen after filtering and sorting because the table height could change

File: utils/rx-page-objects/test/rxPaginate.midway.js (hunk 8 above)
Comment (CITguy) [inline utils/rx-page-objects/test/rxPaginate.midway.js]: @Droogans: the rxPaginate element with the `#rx-paginate-hidden` id isn't present, so this exercise wouldn't be testing anything that `#does-not-exist` is already doing.

**Comment by CITguy** [comment]
@Droogans: any ideas around the codeship failure?

File: utils/rx-page-objects/test/rxPaginate.midway.js (hunk 8 above)
Comment (Droogans) [inline utils/rx-page-objects/test/rxPaginate.midway.js]: The `isPresent` check happens first, and then quits the entire exercise early. The `isDisplayed` check comes immediately after, and then quits early. It's because I can't check a non-present element for `isDisplayed`. That's why there was two sets of tests there.

If there's a way to merge the two tests together (without adding another `rx-debug` example to the html), I'm all for it.

Comment (CITguy) [inline utils/rx-page-objects/test/rxPaginate.midway.js]: Oh, I see. I overlooked the option being passed to the exercise. We'd need another `rx-debug` example to test the `isDisplayed` functionality.

@satyamkotakonda: please revert this change (additional comment coming soon to readd the hidden pagination element).

File: src/elements/Tables/docs/tables.docs.html [hunk 10]
```diff
@@ -138,8 +136,38 @@
...
 
+<h3><rx-permalink>UI-Based Pagination</rx-permalink></h3>
+
+<ul class="list">
+  <li>
+    UI-based pagination, where all items are retrieved at once, and paginated in the UI
+  </li>
+</ul>
+<rx-example name="table.paginate.ui"></rx-example>
+
```
Comment (CITguy) [inline src/elements/Tables/docs/tables.docs.html]: Please add the following below the `table.paginate.ui` example, to satisfy the `#rx-paginate-hidden` midway.

```html
<rx-debug>
  <div ng-controller="rxPaginateUiCtrl">
    <h3>Non-displayed pagination</h3>
    <rx-paginate id="rx-paginate-hidden" page-tracking="pager" ng-hide="true"></rx-paginate>
  </div>
</rx-debug>
```

**Comment by satyamkotakonda** [comment]
@CITguy  @Droogans  Removed midway test case added and please review.
Thanks.

File: src/elements/Tables/scripts/rxPaginate.js [hunk 11]
```diff
@@ -482,10 +482,151 @@
  * `<rx-paginate>` will use the cached values.
  *
  * ## Directives
- * * {@link rxPaginate.directive:rxLoadingOverlay rxLoadingOverlay}
- * * {@link rxPaginate.directive:rxPaginate rxPaginate}
+ * * {@link elements.directive:rxLoadingOverlay rxLoadingOverlay}
+ * * {@link elements.directive:rxPaginate rxPaginate}
```
Comment (CITguy) [inline src/elements/Tables/scripts/rxPaginate.js]: The "Directives" section is unnecessary, considering that this documentation is focused on the `rxPaginate` directive. Please remove the "Directives" section and its associated list.

**Comment by satyamkotakonda** [comment]
@CITguy  Updated changes and please review.
Thanks

**Comment by Droogans** [comment]
I'll see what I can do to get these sortable column test failures working again.

File: src/elements/Tables/scripts/rxPaginate.js [hunk 12]
```diff
@@ -482,10 +479,150 @@
  * `<rx-paginate>` will use the cached values.
  *
  * ## Directives
- * * {@link rxPaginate.directive:rxLoadingOverlay rxLoadingOverlay}
- * * {@link rxPaginate.directive:rxPaginate rxPaginate}
+ * * {@link elements.directive:rxLoadingOverlay rxLoadingOverlay}
```
Comment (CITguy) [inline src/elements/Tables/scripts/rxPaginate.js]: Please remove this link and the "Directives" header as requested.

**Comment by Droogans** [comment]
Codeship failures stem from dumping multiple tables into a single page when before there was a 1:1 relationship to demo pages and testable elements -- e.g., there's more than one sortable column named "Name" now, so everything is clicking it.

Also, the second table has an ugly "Testing Protractor Sorts" column....pretty obvious that's not meant for rx-example.

I'll rip it out and rx-debug it directly into the main demo page.

**Comment by satyamkotakonda** [comment]
@CITguy  Directive link removed and please take a look. 
Thanks.

**Comment by satyamkotakonda** [comment]
@CITguy  Can squash the commit messages.

**Comment by CITguy** [comment]
@satyamkotakonda: don't worry about squashing. It's simple enough that I'll handle it when this PR is ready to merge.

**Comment by cgtaab** [comment]
Design LGTM
